### PR TITLE
PHOENIX-2182

### PR DESCRIPTION
**Pull request by mujtabachohan**

### CODE REVIEW

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/Pherf.java [hunk 1]
```diff
@@ -76,6 +78,8 @@
         options.addOption("d", "debug", false, "Put tool in debug mode");
         options.addOption("stats", false,
                 "Update Phoenix Statistics after data is loaded with -l argument");
+		options.addOption("label", true, "Label a run. Result file name will be suffixed with specified label");
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/Pherf.java]: Is this an override to then dynamically generated label? Also, it would be good if the label was applied as a column value in the result data. That way we can distinguish between runs on analysis.

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/Pherf.java [hunk 2]
```diff
@@ -136,6 +142,8 @@ public Pherf(String[] args) throws Exception {
                 command.getOptionValue("writerThreadSize",
                         properties.getProperty("pherf.default.dataloader.threadpool"));
         properties.setProperty("pherf. default.dataloader.threadpool", writerThreadPoolSize);
+        label = command.getOptionValue("label", null);
+        compareResults = command.getOptionValue("compare", null);
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/Pherf.java]: default to false instead of null

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/Pherf.java [hunk 3]
```diff
@@ -179,6 +188,15 @@ public void run() throws Exception {
                 }
                 return;
             }
+
+            // Compare results and exit  
+			if (null != compareResults) {
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/Pherf.java]: if(!compareResults){}

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/Pherf.java (hunk 2 above)
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/Pherf.java]: Use boolean and set to false as default

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/Pherf.java (hunk 3 above)
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/Pherf.java]: if(!compareResults){}

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/Pherf.java [hunk 4]
```diff
@@ -179,6 +188,15 @@ public void run() throws Exception {
...
+				logger.info("\nStarting to compare results and exiting for " + compareResults);
+				ResultsComparator rc = new ResultsComparator(compareResults);
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/Pherf.java]: Should this be the label?

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/PherfConstants.java [hunk 5]
```diff
@@ -62,6 +62,8 @@
     public static final int MONITOR_FREQUENCY = 5000;
     public static final String MONITOR_FILE_NAME = "STATS_MONITOR";
 
+    public static final String SUMMARY_HTML_FILE_NAME = "summary.html";
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/PherfConstants.java]: Can you put this in the properties file?

Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/PherfConstants.java]: I am trying to make everything here configurable from properties.

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/result/ResultUtil.java [hunk 6]
```diff
@@ -147,9 +147,9 @@ public String getSuffix() {
         if (null == FILE_SUFFIX) {
             Date date = new Date();
             Format formatter = new SimpleDateFormat("YYYY-MM-dd_hh-mm-ss");
-            FILE_SUFFIX = "_" + formatter.format(date);
+            FILE_SUFFIX = formatter.format(date);
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/result/ResultUtil.java]: Should you call setFileSuffix()?

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/PhoenixUtil.java [hunk 7]
```diff
@@ -274,4 +274,34 @@ public void updatePhoenixStats(String tableName, Scenario scenario) throws Excep
...
+     * @param query
+     * @return
+     * @throws SQLException
+     */
+    public String getExplainPlan(Query query) throws SQLException {
+        Connection conn = null;
+        ResultSet rs = null;
+        PreparedStatement statement = null;
+        StringBuilder buf = new StringBuilder();
+        try {
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/PhoenixUtil.java]: You might want to consider using the JdbcSession() to handle all the boiler plate try/catch/finally stuff on the connection. 

Check out testRWWorkload()
https://github.com/apache/phoenix/blob/master/phoenix-pherf/src/it/java/org/apache/phoenix/pherf/DataIngestIT.java

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/ResultsComparator.java [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+import org.apache.phoenix.pherf.PherfConstants;
+import org.apache.phoenix.pherf.result.file.ResultFileDetails;
+
+/**
+ * Compare results based on set threshold and render results as Google Charts
+ */
+public class ResultsComparator {
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/ResultsComparator.java]: Should this implement Comparable? Calling it Comparator is misleading if it's not doing that.

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/ResultsComparator.java [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+
+	private String[] labels;
+	private final Map<String, DataNode> datanodes = new TreeMap<String, DataNode>();
+	private final PherfConstants constants = PherfConstants.create();
+	private final String resultDir = constants.getProperty("pherf.default.results.dir");
+	private final double threshold = Double.parseDouble(constants.getProperty("pherf.default.comparison.threshold"));
+
+	public ResultsComparator(String labels) {
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/ResultsComparator.java]: nit: label should be singular ;)

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/ResultsComparator.java [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+	
+	String[] getLabels() {
+		return labels;
+	}
+
+	void setLabels(String[] labels) {
+		this.labels = labels;
+	}
+
+	void setLabelsAsString(String labels) {
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/ResultsComparator.java]: This is setting as array. Why not just call it setLabels() and let overloading do it's thing?

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/ResultsComparator.java [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+    	}
+    	return true;
+    }
+    
+    /**
+     * Render results as Google charts
+     * @throws FileNotFoundException
+     * @throws UnsupportedEncodingException
+     */
+    private void renderAsGoogleChartsHTML() throws FileNotFoundException, UnsupportedEncodingException {
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/ResultsComparator.java]: Should all this chart stuff go in a util object? It seems unrelated to a Result comparison.

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/ResultsComparator.java [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+			e.printStackTrace();
+		}
+	}
+
+	/**
+	 * Reads aggregate file and convert it to DataNode 
+	 * @param label
+	 * @throws Exception
+	 */
+    private void read(String label) throws Exception {
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/ResultsComparator.java]: Can we use a CSVResultHandler with a ResultFileDetails.CSV_AGGREGATE_PERFORMANCE? It looks like there is a fair amount of redundancy with that and the ResultWriter. It makes sense that you need to wrap the compare logic into this, but result reading and writing should use the existing framework for that.

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/ResultsComparator.java [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+    		this.setTenantId(tenantId);
+    		this.setLabel(label);
+    		this.setRowCount(rowCount);
+    	}
+    	
+		String getExplainPlan() {
+			return explainPlan;
+		}
+		String getExplainPlanAsHTML() {
+			return "</br><font face=arial size=1><b>EXPLAIN PLAN </b>"
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/util/ResultsComparator.java]: Perhaps we should use an html generation library instead of hard coding all this in strings?

Never tried this one, bit looks like it has god template support.
http://freemarker.org/docs/dgui_quickstart_basics.html

File: phoenix-pherf/src/main/java/org/apache/phoenix/pherf/workload/MultithreadedDiffer.java [hunk 14]
```diff
@@ -51,7 +53,7 @@ private void diffQuery() throws Exception {
         Date startDate = Calendar.getInstance().getTime();
         String newCSV = queryVerifier.exportCSV(query);
         boolean verifyResult = queryVerifier.doDiff(query, newCSV);
-        String explainPlan = queryVerifier.getExplainPlan(query);
```
Comment (codymarcel) [inline phoenix-pherf/src/main/java/org/apache/phoenix/pherf/workload/MultithreadedDiffer.java]: Could you add the explainPlan stuff as a different Jira/CL? It might get checked in faster if it's detached from this CL. This one might need a little more love.

**Comment by stoty** [comment]
Already merged.
